conans/test/functional/graph_lock/dynamic_test.py

```diff
@@ -254,3 +255,110 @@ def augment_test_package_requires(self):
         else:
             self.assertEqual(dep["ref"], "dep/0.1")
             self.assertEqual(dep["prev"], "0")
+
+
+class PartialOptionsTest(unittest.TestCase):
+    """
+    When an option is locked in an existing lockfile, and we are using that lockfile to
+    create a new one, and somehow the option is changed there are 2 options:
+    - Allow the non-locked packages to change the value, according to the dependency resolution
+      algorithm. That will produce a different package-id that will be detected and raise
+      as incompatible to the locked one
+    - Force the locked options, that will result in the same package-id. The package attempting
+      to change that option, will not have it set, and can fail later (build time, link time)
+
+    This test implements the 2nd approach, let the lockfile define the options values, not the
+    package recipes
+    """
+    def setUp(self):
+        client = TestClient()
+        client.run("config set general.default_package_id_mode=full_package_mode")
+        client.save({"conanfile.py": GenConanfile().with_option("myoption", [True, False])})
+        client.run("create . LibA/1.0@ -o LibA:myoption=True")
+        self.assertIn("LibA/1.0:d2560ba1787c188a1d7fabeb5f8e012ac53301bb - Build", client.out)
+        client.run("create . LibA/1.0@ -o LibA:myoption=False")
+        self.assertIn("LibA/1.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Build", client.out)
+        self.client = client
+
+    def partial_lock_option_command_line_test(self):
```

Comment:
I am not sure I understand the issue. The lockfile stores all the inputs, the "effective profile" applied to create the lock, including the options, in the same way that it stores the settings, build_requires and environment variables. This stored profile is applied entirely when the lockfile is used. Without it, lockfiles would be a bit incomplete as they will not be enough to reconstruct a dependency graph deterministically.

Reply:
I agree, the `snapshot, not instructions` statement doesn't seem valid, but I think @memsharded comments probably clear it up.  Lockfiles definitely are "instructions", and so if that is unclear maybe we should look to clarify it in the documentation.

Reply:
For me it's also a bit strange the change of behaviour, if you use the `-o` in the command line you end with a `profile_host` in the lockfile that includes the option `[options]\nLibA:myoption=True` but you also have that info in the nodes of the lockfile:

```
   "2": {
    "ref": "LibA/1.0",
    "options": "myoption=True",
    "package_id": "d2560ba1787c188a1d7fabeb5f8e012ac53301bb",
    "prev": "0",
    "context": "host"
   }
```
That makes that that option will affect nodes that are not yet in the lockfile...

Maybe the profile should not change and the -o should only affect to the information stored in the nodes of the lockfile?

Reply:
I know @memsharded said the string value of the profile is effectively serving a "unique key" and part of the "identity" of the lockfile. I think it might be really hard to figure out a way to change that.

Reply:
After talking to @czoido and thinking about it... and some drawing 😄  I think that maybe the _graph+profile_ is easier to manage than the _snapshot_ approach when dealing with lockfiles that contains a subgraph. Here you can find some of my reasoning (probably only clear to me): https://docs.google.com/presentation/d/1G8E5sPfWN04NmJyjagjFPJeQgcPqmTiwTTDkzZySuCo/edit?usp=sharing

I've realized that my main concern is about using a lockfile _smaller_ than the graph: using the lockfile from `LibB` to build the `App`, for example. As the lockfile is also providing the profile, it will affect packages that are outside the lockfile itself. I mean, if we created the lockfile with `conan lock create .... -o *:sahred=True` it will apply the option `shared=True` to every package of the graph, not only those that are into the lockfile. Alternatives:
 * Do not store command line arguments in the _profile_, value will be applied only to packages inside the lockfile.
 * Do not let the user run `conan create ... --lockfile=libb.lock` if the lockfile doesn't cover all the packages. The user will be forced to run `conan lock create ... --lockfile --lockfile-out [config for _extra_ packages]` to build the full lockfile, and then use it for `conan create`.

I hope my concern is clear enough. After changing my mind these are just details... but yes, I would need to read something else in the documentation to clarify it from the very beginning (maybe it is just me, it happens that when you are working on something you stop reading the docs).

Reply:
> Do not let the user run conan create ... --lockfile=libb.lock if the lockfile doesn't cover all the packages. 

``conan create`` and ``conan install`` doesn't allow at all to bring new packages to the graph. All packages involved in those commands will be covered and locked by the lockfiled, otherwise, it will fail

The mechanism to deal with this, is effectively, as you said, to create a new lockfile from the previous. It is possible to compute a new one with a new profile, if the lockfile used as a base was captured without profile (``--base`` argument). It is also possible to remove the locked profile and create a new lockfile without it with the ``--base`` argument, from a full lockfile. So at least with a 2 step process (first remove profile information with ``--base``, then use that base lockfile and provide new profile arguments), this seems doable.